packages/@lwc/engine-server/src/renderer.ts

```diff
@@ -330,23 +338,114 @@ function isConnected(node: HostNode) {
 // synthetic shadow.
 const insertStylesheet = noop as (content: string, target: any) => void;
 
-// Noop on SSR.
-const addEventListener = noop as (
-    target: HostNode,
+function addEventListener(
+    node: N,
     type: string,
     callback: EventListener,
-    options?: AddEventListenerOptions | boolean
-) => void;
+    useCaptureOrOptions?: AddEventListenerOptions | boolean
+): void {
+    if (node.type !== 'element') {
+        return;
+    }
+
+    let savedCallback = callback;
+
+    if (useCaptureOrOptions) {
+        if (useCaptureOrOptions === true || useCaptureOrOptions.capture) {
+            // Capturing listeners aren't meaningful in SSR, since the capture phase
+            // is not supported. However, we don't need to break apps that register
+            // these listeners.
+            return;
+        }
+        if (useCaptureOrOptions.once) {
+            savedCallback = (...args) => {
+                try {
+                    callback(...args);
+                } finally {
+                    removeEventListener(node, type, savedCallback);
+                }
+            };
+        }
+
+        // In the SSR context, `passive` event listeners are not meaningful and are ignored.
+        // Additionally, we do not currently support AbortSignals, so the `signal` option
+        // is also ignored.
+    }
 
-// Noop on SSR.
-const removeEventListener = noop as (
-    target: HostNode,
+    if (!(type in node.eventListeners)) {
+        node.eventListeners[type] = new Set();
+    }
+    node.eventListeners[type].add(savedCallback);
+}
+
+function removeEventListener(
+    node: N,
     type: string,
-    callback: EventListener,
-    options?: AddEventListenerOptions | boolean
-) => void;
+    callback: EventListener
+    // captured listeners aren't supported in SSR, so options are ignored
+): void {
+    if (node.type !== 'element') {
+        return;
+    }
+    const eventListeners = node.eventListeners[type];
```

Comment:
Nitpick: I find it a bit confusing that `eventListeners` can refer to both a `Record<string, Set>` and the `Set`. Maybe we can call this variable `listenersSet` or something? Or rename `node.eventListeners`?